javascript/ql/src/semmle/javascript/frameworks/Next.qll

```diff
@@ -0,0 +1,240 @@
+   */
+  DataFlow::FunctionNode getInitialProps(Module pageModule) {
+    pageModule = getAPagesModule() and
+    (
+      result =
+        pageModule
+            .getAnExportedValue("default")
+            .getAFunctionValue()
+            .getAPropertyWrite("getInitialProps")
+            .getRhs()
+            .getAFunctionValue()
+      or
+      result =
+        pageModule
+            .getAnExportedValue("default")
+            .getALocalSource()
+            .getAstNode()
+            .(ReactComponent)
+            .getStaticMethod("getInitialProps")
+            .flow()
+    )
+  }
+
+  /**
+   * Gets a reference to a `props` object computed by the Next.js server.
+   * This `props` object is both used both by the server and client to render the page.
+   */
+  DataFlow::Node getAPropsSource(Module pageModule) {
+    pageModule = getAPagesModule() and
+    (
+      result =
+        [getStaticPropsFunction(pageModule), getServerSidePropsFunction(pageModule)]
+            .getAReturn()
+            .getALocalSource()
+            .getAPropertyWrite("props")
+            .getRhs()
+      or
+      result = getInitialProps(pageModule).getAReturn()
+    )
+  }
+
+  /**
+   * A step modelling the flow from the server-computed props object to the default exported function that renders the page.
+   */
+  class NextJSStaticPropsStep extends DataFlow::AdditionalFlowStep, DataFlow::FunctionNode {
+    Module pageModule;
+
+    NextJSStaticPropsStep() {
+      pageModule = getAPagesModule() and
+      this = pageModule.getAnExportedValue("default").getAFunctionValue()
+    }
+
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      pred = getAPropsSource(pageModule) and
+      succ = this.getParameter(0)
+    }
+  }
+
+  /**
+   * A step modelling the flow from the server-computed props object to the default exported React component that renders the page.
+   */
+  class NextJSStaticReactComponentPropsStep extends DataFlow::AdditionalFlowStep,
+    DataFlow::ValueNode {
+    Module pageModule;
+    ReactComponent component;
+
+    NextJSStaticReactComponentPropsStep() {
+      pageModule = getAPagesModule() and
+      this.getAstNode() = component and
+      this = pageModule.getAnExportedValue("default").getALocalSource()
+    }
+
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      pred = getAPropsSource(pageModule) and
+      succ = component.getADirectPropsAccess()
+    }
+  }
+
+  /**
+   * A Next.js function that is exected on the server for every request, seen as a routehandler.
+   */
+  class NextHttpRouteHandler extends HTTP::Servers::StandardRouteHandler, DataFlow::FunctionNode {
+    Module pageModule;
+
+    NextHttpRouteHandler() {
+      this = getServerSidePropsFunction(pageModule) or this = getInitialProps(pageModule)
+    }
+  }
+
+  /**
+   * A NodeJS HTTP request object in a Next.js page.
+   */
+  class NextHttpRequestSource extends NodeJSLib::RequestSource {
```

Comment:
Is there a reason these are modelled as Node.js requests, while requests in the `api` folder are modelled as Express requests?

Reply:
Yes.  

They are both Node.js HTTP requests/response objects.  
But the `res` object in API Routes includes a set of Express.js-like methods (doc: https://nextjs.org/docs/api-routes/response-helpers).  
I therefore choose to model that part as an Express route-handler.